alf/algorithms/algorithm.py

```diff
@@ -544,10 +569,18 @@ def _add_params_to_optimizer(params, opt):
             assert id(child) != id(self), "Child should not be self"
             if isinstance(child, Algorithm):
                 params, child_handled = child._setup_optimizers_(param_to_name)
-                for m in child_handled:
-                    assert m not in handled, duplicate_error % param_to_name.get(
-                        m)
-                    handled[m] = 1
+                if child.force_params_visible_to_parent:
+                    merged = set(params)
+                    for p in child_handled:
+                        if p in merged:
```

Comment:
When will this happen? By definition, `params` are not handled by another child optimizer, while `child_handled` are those handled. Should these two be disjoint always? 
(And it seems unnecessary to introduce `merged`? Just `if p in params`.

Reply:
I think they should be disjoint, but not 100% certain - that is why I added this logic to ensure correctness.
> And it seems unnecessary to introduce merged? Just if p in params.
Agree that `p in params` may not be much slower in most case. I am being conservative here to use a hashtable (i.e. `set`) because the number of parameters can potentially be high (i.e. > 100). `_setup_optimizers_` is called multiple times every iteration for now, and that is why I am a bit sensitive on performance here.